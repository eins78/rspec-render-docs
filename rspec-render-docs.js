#!/usr/bin/env node
import fs from 'fs'
import url from 'url'
import React from 'react'
import f from 'active-lodash'
import { renderToStaticMarkup as renderReact } from 'react-dom/server'
import SyntaxHighlighter from 'react-syntax-highlighter'
import ReactMarkdown from 'react-markdown'
import { githubGist } from 'react-syntax-highlighter/dist/styles'
const slugify = (str) => f.kebabCase(f.deburr(str))
process.on('uncaughtException', (e) => { console.error('Error!', e); process.exit(1) })

const pageTitle = 'Madek Features'
const versionLinkPath = 'https://ci.zhdk.ch/cider-ci/ui/workspace/trees/'
const sourceCodeUrlPath = 'https://github.com/Madek/madek-webapp/'
const showUselessSteps = true // show 1 line of step - title will be the same (currently all)

const rspecData = JSON.parse(fs.readFileSync(process.argv[2]))
if (!f.present(f.get(rspecData, 'examples'))) throw new Error('No Data!')

// NOTE: only supports 2 levels of nesting, so need to get full_description from example_group (but still only 1. line - title!)
const chapters = f(rspecData.examples)
  // TODO: .map((e) => f.merge(e, {slug: slugify()})) â€¦
  .sortBy((ex) => f.get(ex, 'example_group.scoped_id'))
  .groupBy((ex) => f.first(f.get(ex, 'example_group.full_description').split('\n')))
  .pairs()
  .sortBy(([key]) => (key || '').toLowerCase()) // ascii order, but ignore case
  .groupBy(([key]) => f.first(key.split(':')).trim())
  .pairs()
  .value()

const RspecStory = (rspecData) => {
  const gitCommit = rspecData.git_commit
  const gitTree = rspecData.git_tree
  const gitRef = gitCommit || 'master'
  const versionLink = gitTree && url.resolve(versionLinkPath, `./${gitTree}`)
  const sourceCodeLink = url.resolve(sourceCodeUrlPath + './blob/' + gitRef + '/', '.')

  const Document = ({children}) => (
    <html>
      <head>
        <meta charSet='utf-8' />
        <title>{pageTitle}</title>
        {/* steal a stolen version of github's markdown doc style: */}
        <link rel='stylesheet' href='https://cdn.rawgit.com/sindresorhus/github-markdown-css/gh-pages/github-markdown.css' />
        <style>{'.markdown-body { box-sizing: border-box; min-width: 200px; max-width: 980px; margin: 0 auto; padding: 45px;}'}</style>
      </head>
      <body className='markdown-body'>{children}</body>
    </html>
  )

  const CodeBlock = ({l, children}) => (
    <SyntaxHighlighter language={l} style={githubGist}
      customStyle={{background: null}} children={children || ''} />
  )

  // inline-only markdown
  const Md = ({children}) => (
    <ReactMarkdown source={children} escapeHtml
      unwrapDisallowed allowedTypes={['Text', 'Strong', 'Emph', 'Code']} />
  )

  // document-internal link from text, self-links when isHeading
  const NavLink = ({text, isHeading, children}) => {
    const slug = slugify(text || children)
    return <a href={'#' + slug} id={isHeading ? slug : null}>{children}</a>
  }

  const TableOfContents = (chapters) => (
    <ul>
      {f.map(chapters, ([title, sections]) => [
        <li><NavLink text={title}><Md>{title}</Md></NavLink></li>,
        <ul>
          {f.map(sections, ([title, examples]) => [
            <li><NavLink text={title}><Md>{title}</Md></NavLink></li>,
            <ul>
              {f.map(examples, (example, n) => (
                <li key={title + example.description + n}>
                  <NavLink text={`${title}-${example.description}`}>
                    <Md>{example.description}</Md>
                  </NavLink>
                </li>
              ))}
            </ul>
          ])}
        </ul>
      ])}
    </ul>
  )

  const ExampleSection = ({title, example}) => {
    const {description, file_path, source} = example
    const [lineNumber, sourceBlock] = [example.line_number, example.source_block]
    const fileName = f.present(file_path) &&
      file_path.replace(/^\.\//, '') +
      (lineNumber ? ':' + lineNumber : '')

    const githubLink = fileName && f.isNumber(lineNumber) &&
      url.resolve(sourceCodeLink, file_path) +
      (lineNumber ? '#L' + lineNumber : '')

    const lines = description.split('\n')
    const steps = lines.slice(1).join('\n')
    const showSteps = steps || showUselessSteps

    return (<div>
      {showSteps && <h6>Steps</h6>}
      {showSteps && <ReactMarkdown source={steps || description} />}
      <h6>
        {'Source '}
        {fileName && <a href={githubLink} style={{float: 'right'}}><kbd>{fileName}</kbd></a>}
      </h6>
      <CodeBlock l='ruby'>{sourceBlock}</CodeBlock>
      <div className='hook-debug' style={{display: 'none'}}>
        <pre><code>{source}</code></pre>
        <CodeBlock l='json'>{JSON.stringify(example, 0, 2)}</CodeBlock>
      </div>
    </div>)
  }

  const Chapters = (chapters) => (
    <article>
      {f.map(chapters, ([title, sections]) =>
        <section key={title}>
          <h2 style={{display: 'none'}}><NavLink text={title}><Md>{title}</Md></NavLink></h2>

          {f.map(sections, ([title, examples], index) => [
            <h3><NavLink isHeading text={title}><Md>{title}</Md></NavLink></h3>,
            f.map(examples, (example) => [
              <h4>
                <NavLink isHeading text={`${title}-${example.description}`}>
                  <Md>{example.description}</Md>
                </NavLink>
              </h4>,
              ExampleSection({title, example}),
              ((index < sections.length) && <hr />)
            ])
          ])}
        </section>
      )}
    </article>
  )

  return (<Document>
    <h1>
      {pageTitle + ' '}
      {versionLink &&
        <small style={{float: 'right'}}>
          <span style={{top: '0.25em', position: 'relative'}}>ðŸŒ´ </span>
          <a href={versionLink}><code>{gitTree.slice(0, 8)}</code></a></small>
      }
    </h1>
    <p>generated from RSpec test output</p>
    <h2>Overview</h2>
    {TableOfContents(chapters)}
    <hr />
    {Chapters(chapters)}
  </Document>)
}

process.stdout.write('<!DOCTYPE html>' + renderReact(RspecStory(rspecData)))
