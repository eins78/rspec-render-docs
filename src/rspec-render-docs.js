#!/usr/bin/env node

// script to render (HTML) docs from rspec data

import fs from 'fs'
import f from 'lodash'
import { buffer as getStdin } from 'get-stdin'
import { renderToStaticMarkup as renderReact } from 'react-dom/server'
import RspecStory from './ui/RspecStory'

const exitOnErr = e => {
  console.error('Error!', e)
  process.exit(1)
}
process.on('uncaughtException', exitOnErr)

// user options
const opts = {
  pageTitle: 'Madek Features',
  versionLinkPath: 'https://ci.zhdk.ch/cider-ci/ui/workspace/trees/',
  sourceCodeUrlPath: 'https://github.com/Madek/madek-webapp/',
  // show 1 line of step - title will be the same (currently all)
  showUselessSteps: false
}

// build chapters for view
// NOTE: only supports 2 levels of nesting, so need to get full_description
//       from example_group (but still only 1. line - title!)
export const buildChapters = ({ examples }) => f
  .chain(examples)
  .sortBy(e => f.get(e, 'example_group.scoped_id'))
  .groupBy(e => f.first(f.get(e, 'example_group.full_description').split('\n')))
  .toPairs()
  .sortBy(([ key ]) => (key || '').toLowerCase())
  .groupBy(([ key ]) => f.first(key.split(':')).trim())
  .toPairs()
  .value()

const renderFromJSONString = str => {
  const rspecData = JSON.parse(str)
  if (f.isEmpty(f.get(rspecData, 'examples'))) throw new Error('No Data!')

  // build data for view
  const chapters = buildChapters(rspecData)
  const config = {
    ...opts,
    gitTree: rspecData.git_tree,
    gitCommit: rspecData.git_commit,
    summary: rspecData.summary
  }

  // output rendered view
  process.stdout.write(
    '<!DOCTYPE html>' + renderReact(RspecStory({ chapters, config }))
  )
}

// read JSON data from filename argument or stdin & render
if (process.argv[2]) {
  fs.readFile(process.argv[2], (err, res) => {
    if (err) {
      throw new Error(err)
    }
    renderFromJSONString(res)
  })
} else {
  getStdin().then(str => renderFromJSONString(str)).catch(exitOnErr)
}
