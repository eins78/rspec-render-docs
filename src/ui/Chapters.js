import React from 'react'
import map from 'lodash/map'

import NavLink from './NavLink'
import { Md } from './Markdown'
import ExampleSection from './ExampleSection'

const Chapters = ({ chapters, sourceCodeLink, showUselessSteps }) => (
  <article>
    {map(chapters, ([ title, sections ]) => (
      <section key={title}>
        <h2 style={{ display: 'none' }}>
          <NavLink text={title}><Md>{title}</Md></NavLink>
        </h2>
        {map(sections, ([ title, examples ], index) => [
          <h3>
            <NavLink isHeading text={title}><Md>{title}</Md></NavLink>
          </h3>,
          map(examples, example => [
            <h4>
              <NavLink isHeading text={`${title}-${example.description}`}>
                <Md>{example.description}</Md>
              </NavLink>
            </h4>,
            ExampleSection({
              title,
              example,
              config: { sourceCodeLink, showUselessSteps }
            }),
            index < sections.length && <hr />
          ])
        ])}
      </section>
      ))}
  </article>
)

export default Chapters
