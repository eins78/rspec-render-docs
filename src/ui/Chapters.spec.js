/* eslint-env jest */
// import React from 'react'
import renderer from 'react-test-renderer'

import Chapters from './Chapters'
import { buildChapters } from '../rspec-render-docs'
import fixture from '../../spec/fixtures/rspec-features.json'
import config from '../../spec/fixtures/config.json'
const chapters = buildChapters(fixture)

it('renders', () => {
  const component = renderer.create(Chapters({ chapters, ...config }))
  expect(component.toJSON()).toMatchSnapshot()
})
