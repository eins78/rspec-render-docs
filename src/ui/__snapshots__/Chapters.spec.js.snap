exports[`test renders 1`] = `
<article>
  <section>
    <h2
      style={
        Object {
          "display": "none",
        }
      }>
      <a
        href="#rspec-section"
        id={null}>
        <div>
          Rspec
          _
          Section
        </div>
      </a>
    </h2>
    <h3>
      <a
        href="#rspec-section"
        id="rspec-section">
        <div>
          Rspec
          _
          Section
        </div>
      </a>
    </h3>
    <h4>
      <a
        href="#rspec-section-example-title"
        id="rspec-section-example-title">
        <div>
          Example
          _
          Title
        </div>
      </a>
    </h4>
    <div>
      <h6>
        Steps
      </h6>
      <div>
        <p>
          Example_Title
        </p>
      </div>
      <h6>
        Source 
        <a
          href="https://code.example.com/repo/files/spec/features/example_spec.rb#L13"
          style={
            Object {
              "float": "right",
            }
          }>
          <kbd>
            spec/features/example_spec.rb:13
          </kbd>
        </a>
      </h6>
      <pre
        style={
          Object {
            "background": null,
            "color": "#333333",
            "display": "block",
            "overflowX": "auto",
            "padding": "0.5em",
          }
        }>
        <code>
          expect(
          <span
            style={Object {}}>
            1
          </span>
          )
                .to eq 
          <span
            style={Object {}}>
            1
          </span>
        </code>
      </pre>
      <div
        className="hook-debug"
        style={
          Object {
            "display": "none",
          }
        }>
        <pre
          style={
            Object {
              "background": null,
              "color": "#333333",
              "display": "block",
              "overflowX": "auto",
              "padding": "0.5em",
            }
          }>
          <code>
              example 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \'Example_Title\'
            </span>
             
            <span
              style={
                Object {
                  "color": "#a71d5d",
                }
              }>
              do
            </span>
            
                expect(
            <span
              style={Object {}}>
              1
            </span>
            )
                  .to eq 
            <span
              style={Object {}}>
              1
            </span>
            
            
              
            <span
              style={
                Object {
                  "color": "#a71d5d",
                }
              }>
              end
            </span>
            
            
          </code>
        </pre>
        <pre
          style={
            Object {
              "background": null,
              "color": "#333333",
              "display": "block",
              "overflowX": "auto",
              "padding": "0.5em",
            }
          }>
          <code>
            {
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Example_Title\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"full_description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section Example_Title\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"status\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"passed\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"file_path\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"./spec/features/example_spec.rb\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"line_number\"
            </span>
            : 
            <span
              style={Object {}}>
              13
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"run_time\"
            </span>
            : 
            <span
              style={Object {}}>
              0.00001
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"pending_message\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#0086b3",
                }
              }>
              null
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"scoped_id\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"1:1\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"type\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"feature\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"example_group\"
            </span>
            : {
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"scoped_id\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"1\"
            </span>
            ,
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section\"
            </span>
            ,
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"full_description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section\"
            </span>
            
              },
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"source\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"  example \'Example_Title\' do\\n    expect(1)\\n      .to eq 1\\n\\n  end\\n\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"source_block\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"expect(1)\\n      .to eq 1\"
            </span>
            
            }
          </code>
        </pre>
      </div>
    </div>
    <hr />
    <h4>
      <a
        href="#rspec-section-example-title-2"
        id="rspec-section-example-title-2">
        <div>
          Example
          _
          Title
          _
          2
        </div>
      </a>
    </h4>
    <div>
      <h6>
        Steps
      </h6>
      <div>
        <p>
          Example_Title_2
        </p>
      </div>
      <h6>
        Source 
        <a
          href="https://code.example.com/repo/files/spec/features/example_spec.rb#L26"
          style={
            Object {
              "float": "right",
            }
          }>
          <kbd>
            spec/features/example_spec.rb:26
          </kbd>
        </a>
      </h6>
      <pre
        style={
          Object {
            "background": null,
            "color": "#333333",
            "display": "block",
            "overflowX": "auto",
            "padding": "0.5em",
          }
        }>
        <code>
          expect(
          <span
            style={Object {}}>
            1
          </span>
          )
                .to eq 
          <span
            style={Object {}}>
            2
          </span>
        </code>
      </pre>
      <div
        className="hook-debug"
        style={
          Object {
            "display": "none",
          }
        }>
        <pre
          style={
            Object {
              "background": null,
              "color": "#333333",
              "display": "block",
              "overflowX": "auto",
              "padding": "0.5em",
            }
          }>
          <code>
              pending 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \'Example_Title\'
            </span>
             
            <span
              style={
                Object {
                  "color": "#a71d5d",
                }
              }>
              do
            </span>
            
                expect(
            <span
              style={Object {}}>
              1
            </span>
            )
                  .to eq 
            <span
              style={Object {}}>
              2
            </span>
            
            
              
            <span
              style={
                Object {
                  "color": "#a71d5d",
                }
              }>
              end
            </span>
            
            
          </code>
        </pre>
        <pre
          style={
            Object {
              "background": null,
              "color": "#333333",
              "display": "block",
              "overflowX": "auto",
              "padding": "0.5em",
            }
          }>
          <code>
            {
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Example_Title_2\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"full_description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section Example_Title_2\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"status\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"pending\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"file_path\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"./spec/features/example_spec.rb\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"line_number\"
            </span>
            : 
            <span
              style={Object {}}>
              26
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"run_time\"
            </span>
            : 
            <span
              style={Object {}}>
              0.00001
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"pending_message\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#0086b3",
                }
              }>
              null
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"scoped_id\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"1:2\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"type\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"feature\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"example_group\"
            </span>
            : {
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"scoped_id\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"1\"
            </span>
            ,
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section\"
            </span>
            ,
                
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"full_description\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"Rspec_Section\"
            </span>
            
              },
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"source\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"  pending \'Example_Title\' do\\n    expect(1)\\n      .to eq 2\\n\\n  end\\n\"
            </span>
            ,
              
            <span
              style={
                Object {
                  "color": "#795da3",
                }
              }>
              \"source_block\"
            </span>
            : 
            <span
              style={
                Object {
                  "color": "#df5000",
                }
              }>
              \"expect(1)\\n      .to eq 2\"
            </span>
            
            }
          </code>
        </pre>
      </div>
    </div>
    <hr />
  </section>
</article>
`;
